File: lib/models/continued_processing_task_options.dart
/// Options for BGContinuedProcessingTask (iOS 26+).
///
/// This allows long-running user-initiated tasks to continue in the background
/// with Live Activity UI showing progress.
class ContinuedProcessingTaskOptions {
  /// The localized title displayed to the user.
  final String title;

  /// The localized subtitle displayed to the user.
  final String subtitle;

  /// Whether the task requires GPU access.
  ///
  /// Requires the `com.apple.developer.background-tasks.continued-processing.gpu`
  /// entitlement. Not all devices support GPU in background.
  final bool requiresGPU;

  /// The submission strategy for the task.
  ///
  /// - [ContinuedProcessingSubmissionStrategy.fail]: Fail immediately if system
  ///   cannot run the task.
  /// - [ContinuedProcessingSubmissionStrategy.queue]: Queue the task if system
  ///   is busy. Queued tasks are cancelled when app is removed from app switcher.
  final ContinuedProcessingSubmissionStrategy submissionStrategy;

  /// Constructs an instance of [ContinuedProcessingTaskOptions].
  const ContinuedProcessingTaskOptions({
    required this.title,
    required this.subtitle,
    this.requiresGPU = false,
    this.submissionStrategy = ContinuedProcessingSubmissionStrategy.queue,
  });

  /// Returns the data fields of [ContinuedProcessingTaskOptions] in JSON format.
  Map<String, dynamic> toJson() {
    return {
      'continuedProcessingEnabled': true,
      'continuedProcessingTitle': title,
      'continuedProcessingSubtitle': subtitle,
      'continuedProcessingRequiresGPU': requiresGPU,
      'continuedProcessingStrategy': submissionStrategy.value,
    };
  }
}

/// Submission strategy for BGContinuedProcessingTask.
enum ContinuedProcessingSubmissionStrategy {
  /// Fail the submission if there is no room for the task request,
  /// or if the system is under substantial load.
  fail('fail'),

  /// Add the request to a queue if there is no room for the task
  /// or if the system is under substantial load.
  ///
  /// Queued tasks will be cancelled when the user removes your app
  /// from the app switcher.
  queue('queue');

  const ContinuedProcessingSubmissionStrategy(this.value);

  final String value;
}

File: ios/Classes/models/ContinuedProcessingTaskOptions.swift
//
//  ContinuedProcessingTaskOptions.swift
//  flutter_foreground_task
//
//  Created for BGContinuedProcessingTask support (iOS 26+)
//

import Foundation

struct ContinuedProcessingTaskOptions {
  let enabled: Bool
  let title: String
  let subtitle: String
  let requiresGPU: Bool
  let submissionStrategy: SubmissionStrategy

  enum SubmissionStrategy: String {
    case fail = "fail"
    case queue = "queue"
  }

  static func getData() -> ContinuedProcessingTaskOptions? {
    let prefs = UserDefaults.standard

    guard prefs.bool(forKey: CONTINUED_PROCESSING_ENABLED) else {
      return nil
    }

    let title = prefs.string(forKey: CONTINUED_PROCESSING_TITLE) ?? "Processing"
    let subtitle = prefs.string(forKey: CONTINUED_PROCESSING_SUBTITLE) ?? "Task in progress"
    let requiresGPU = prefs.bool(forKey: CONTINUED_PROCESSING_REQUIRES_GPU)
    let strategyString = prefs.string(forKey: CONTINUED_PROCESSING_STRATEGY) ?? "queue"
    let strategy = SubmissionStrategy(rawValue: strategyString) ?? .queue

    return ContinuedProcessingTaskOptions(
      enabled: true,
      title: title,
      subtitle: subtitle,
      requiresGPU: requiresGPU,
      submissionStrategy: strategy
    )
  }

  static func setData(args: Dictionary<String, Any>) {
    let prefs = UserDefaults.standard

    if let enabled = args[CONTINUED_PROCESSING_ENABLED] as? Bool {
      prefs.set(enabled, forKey: CONTINUED_PROCESSING_ENABLED)
    }

    if let title = args[CONTINUED_PROCESSING_TITLE] as? String {
      prefs.set(title, forKey: CONTINUED_PROCESSING_TITLE)
    }

    if let subtitle = args[CONTINUED_PROCESSING_SUBTITLE] as? String {
      prefs.set(subtitle, forKey: CONTINUED_PROCESSING_SUBTITLE)
    }

    if let requiresGPU = args[CONTINUED_PROCESSING_REQUIRES_GPU] as? Bool {
      prefs.set(requiresGPU, forKey: CONTINUED_PROCESSING_REQUIRES_GPU)
    }

    if let strategy = args[CONTINUED_PROCESSING_STRATEGY] as? String {
      prefs.set(strategy, forKey: CONTINUED_PROCESSING_STRATEGY)
    }
  }

  static func updateData(args: Dictionary<String, Any>) {
    setData(args: args)
  }

  static func clearData() {
    let prefs = UserDefaults.standard
    prefs.removeObject(forKey: CONTINUED_PROCESSING_ENABLED)
    prefs.removeObject(forKey: CONTINUED_PROCESSING_TITLE)
    prefs.removeObject(forKey: CONTINUED_PROCESSING_SUBTITLE)
    prefs.removeObject(forKey: CONTINUED_PROCESSING_REQUIRES_GPU)
    prefs.removeObject(forKey: CONTINUED_PROCESSING_STRATEGY)
  }
}
